doc/anaconda.md

```diff
@@ -0,0 +1,73 @@
+Cockpit Storage in Anaconda Mode
+================================
+
+Anaconda (the OS Installer) can open the Cockpit "storaged" for
+advanced setup of the target storage devices. When this is done,
+storaged is in a special "Anaconda mode" and behaves significantly
+different.
+
+In essence, the storaged page restricts itself to working with the
+target environment. It will hide the real root filesystem (on the USB
+stick that the Live environment was booted from), but let the user
+create a "fake" root filesystem on some block device.
+
+Entering Anaconda mode
+----------------------
+
+The "storaged" page is put into Anaconda mode by changing its
+manifest, probably via the /etc/cockpit/storaged.override.json file.
+
+Anaconda mode is activated when the manifest contains a "anaconda"
+entry. All further configuration is done via the value of that entry.
+For example, you could put this into /etc/cockpit/storaged.override.json:
+
+```
+{
+  "anaconda": {
+  }
+}
+```
+
+Ignoring storage devices
+------------------------
+
+Anaconda needs to tell the storaged page which devices can not be used
+to install the OS on. This is done with the "ignore_devices" entry,
+which is a array of strings.
+
+```
+{
+  "anaconda": {
+    "ignore_devices": [ "/dev/sda" ]
+  }
+}
+```
+
+Entries in that array can refer to block devices, LVM2 volume groups
+(/dev/vgroup-name/), and Stratis pools (/dev/stratis/pool-name/).
+
+Mount point prefix
+------------------
+
+The storaged page can be put into a kind of "chroot" environment by
+giving it a mount point prefix like so:
+
+```
+{
+  "anaconda": {
+    "mount_point_prefix": "/sysroot"
+  }
+}
+```
+
+This works at the UI level: filesystems that have mount points outside
+of "/sysroot" are hidden from the user, and when letting the user work
+with mount points below "/sysroot", the "/sysroot" prefix is
+omitted. So when the user says to create a filesystem on "/var", they
+are actually creating one on "/sysroot/var".
+
+However, Cockpit (via UDisks2) will still write the new mount point
+configuration into the real /etc/fstab (_not_ /sysroot/etc/fstab).
+
+Anaconda will read the real /etc/fstab and reverse engineer the
```

Comment:
@vojtechtrefny @VladimirSlavik @poncovka I think Blivet has some support for fstab parsing, right ? I just wonder what else could be there in fstab & if we can't get into a situation where we can't tell what we want to forward to the target system & what now.

Reply:
Blivet doesn't have support for parsing fstab yet, but it is planned. But I think it would be better to not write to fstab -- in the "anaconda" mode Cockpit could simply write the mount points to a custom configuration file, all we need in Anaconda is a simple device -> mountpoint "mapping", we don't need the full fstab entry. We could avoid a lot of potential issues by not writing to fstab on the live image (which is not empty by the way).

Reply:
We (now speaking as a udisks developer) could also add a new option/type to `AddConfigurationItem` to be able to write to a fstab in a different location than the default `/etc/fstab`.

Reply:
> We (now speaking as a udisks developer) could also add a new option/type to [AddConfigurationItem](http://storaged.org/doc/udisks2-api/latest/gdbus-org.freedesktop.UDisks2.Block.html#gdbus-method-org-freedesktop-UDisks2-Block.AddConfigurationItem) to be able to write to a fstab in a different location than the default `/etc/fstab`.

I think that would only help if UDisks2 would also watch and read that different location. If modifying the real /etc/fstab is not acceptable, I think I would rather change Cockpit Storage to maintain its own "virtual fstab" (in Browser storage?) and use that instead of the UDisks2 API.

Reply:
> We could avoid a lot of potential issues by not writing to fstab on the live image (which is not empty by the way).

What issues do you see?

I can't think of any, tbh.  Cockpit Storage will of course not corrupt /etc/fstab, if that is one of your concerns.

Reply:
> I just wonder what else could be there in fstab & if we can't get into a situation where we can't tell what we want to forward to the target system & what now.

The argument for writing into /etc/fstab is that this needs no code changes to Cockpit Storage. The main challenge I see is that Anaconda needs more code. :-)

There must be Python code for parsing fstab files, but Anaconda probably also needs to look into udev to find the block device that holds a filesystem with a given UUID.

If that is too problematic, Cockpit can also, in addition to /etc/fstab, write a separate file with new mountpoints. Then /etc/fstab would only be changed to make live easier for Cockpit.

Reply:
> all we need in Anaconda is a simple device -> mountpoint "mapping"

Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?

Reply:
Cockpit now stores a mountpoint -> device map for Anaconda.

Reply:
> I think that would only help if UDisks2 would also watch and read that different location.

I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.

> What issues do you see?

My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

I am also not sure how well Cockpit (or UDisks to be more precise) handles btrfs subvolumes in fstab, but that's a different story.

> There must be Python code for parsing fstab files

Anaconda has some functions for parsing fstab and we are working on more generic support in blivet (see https://github.com/storaged-project/blivet/pull/1119)

> but Anaconda probably also needs to look into udev to find the block device that holds a filesystem with a given UUID.

Blivet has this covered.

> Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?

Yes, we (blivet) can identify a device by name, uuid, all the `/dev` symlinks, basically everything.

```
In [2]: b.devicetree.resolve_device("/dev/dm-2")
Out[2]: 
LVMLogicalVolumeDevice instance (0x7f53ec3957d0) --
  name = fedora_aida-root  status = True  id = 78
  ...
```

Reply:
> > I think that would only help if UDisks2 would also watch and read that different location.
> 
> I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.
> 
> > What issues do you see?
> 
> My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

Does Anaconda already look into fstab, or are you referring to Anaconda getting confused during my proposed 'reverse engineering'? If the latter, I think this becomes moot now that Cockpit provides Anaconda with a separate mount-point -> device mapping.

> I am also not sure how well Cockpit (or UDisks to be more precise) handles btrfs subvolumes in fstab, but that's a different story.

Once Cockpit supports btrfs, it will handle the subvolumes correctly, see #16408. But currently it is indeed pretty bad at it.

> > Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?
> 
> Yes, we (blivet) can identify a device by name, uuid, all the `/dev` symlinks, basically everything.

Nice!

Reply:
> > I think that would only help if UDisks2 would also watch and read that different location.
> 
> I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.

@tbzatek would not be so happy with this I'm afraid.

> > What issues do you see?
> 
> My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

I have no idea how entries from multiple sources will be visualized (on the D-Bus interface). Besides, the whole RFE seems a little off the UDisks project target.

However, if something like this is about to be needed, we need to think in bigger picture. There are various files that we watch and/or modify. In some cases where these files are consumed by other libraries we pull in - a problem with different prefix in the build time sometimes causes unexpected issues.

Which brings me into another question - should this be considered with (not actually planned yet) mount namespace support? Would this be interesting for your use case?

Feel free to detach this conversation to a separate ticket.